actors/market/src/lib.rs

```diff
@@ -1038,29 +1184,20 @@ fn get_proposals<BS: Blockstore>(
         if !seen_deal_ids.insert(deal_id) {
             return Err(actor_error!(illegal_argument, "duplicate deal ID {} in sector", deal_id));
         }
-        let proposal = proposal_array
-            .get(*deal_id)
-            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load deal")?
-            .ok_or_else(|| {
-                if *deal_id < next_id {
-                    ActorError::unchecked(EX_DEAL_EXPIRED, format!("deal {} expired", deal_id))
-                } else {
-                    actor_error!(not_found, "no such deal {}", deal_id)
-                }
-            })?;
-        proposals.push((*deal_id, proposal.clone()));
+        let proposal = get_proposal(proposal_array, *deal_id, next_id)?;
+        proposals.push((*deal_id, proposal));
     }
     Ok(proposals)
 }
 
 fn compute_data_commitment(
     rt: &impl Runtime,
-    proposals: &[(DealID, DealProposal)],
+    proposals: &mut dyn Iterator<Item = &DealProposal>,
```

Comment:
why `dyn Iterator` instead of `impl Iterator`?

Reply:
Thanks.

> why dyn Iterator instead of impl Iterator?

Because Rust is complicated and confusing.